### ISSUE TITLE
This is unusable as an API

### ISSUE THREAD

**Issue by ameesme** [Wickedlasers/libLaserdockCore#12]
There's barely any documentation, the compiled binary is massive and there's no integration examples. I am beyond lost on how this can be used to be properly integrated in other software.

**Comment by gavrushkin** [comment]
@ameesme yep, binary is quite massive and depends on Qt library and some other libraries too. There is a simple example, did you check it? https://github.com/Wickedlasers/libLaserdockCore/tree/master/example

Sorry for the lack of documentation, we are trying to document some parts directly in the code.

If you have any questions please feel free to ask here

**Comment by sebleedelisle** [comment]
Hey Sergey, I hope you're doing well... sorry to hijack the thread, but I'm
in a similar position - unfortunately the QT dependencies make it
unworkable for me as well, so I'm going to have to implement the new
network protocol from scratch. Right now this is a pretty difficult task as
I have to work out the protocol from your source code. Is there any way you
could document the protocol, or at least help me through the process? I'd
be very grateful, thanks!

Seb

**Comment by dlernstrom** [comment]
Hey folks

Good dialog here.  @gavrushkin has done quite an impressive job with demonstrating the capabilities of the LaserDock.  I also tried to connect to it directly a couple of years ago and found myself in the same position you were at picking apart the source code from laserdockCore.  My results for speaking to LaserDock over USB with Python are here: https://github.com/dlernstrom/laserdock_python.  The code there is free for the taking.

Admittedly, I got distracted with other things and put down the project a couple of years ago, but that doesn't mean that my efforts wouldn't be a source of insight to you if you are trying to pick out the API from the source code.  My work was in Python and allowed a RPi device to make the LaserDock hum.

Inside my laser_dock.py file, you'll see my best effort at understanding the various parameters and such for the commands.  I wish I could guarantee that they were all accurately depicted.  I can say that my project ran fine, though, so I at least understood parts of it!  :)

Cheers,
David

**Comment by ameesme** [comment]
Hi David,

Thanks for chipping in. Unfortunately, USB support is not the issue here. With the introduction of the new Lasercube, WiFi-support was added. Currently, `libLaserdockCore` is the only project actually supporting network-controls of Lasercube.

Cheers!

**Comment by dlernstrom** [comment]
Ah, got it - my bad.  Didn't realize you were connecting over WiFi.

I'm surprised that the swap out of the connection protocol meant that none of the terminology, methodology, etc., is usable anymore and you have to start from scratch.  Good luck in your project!

**Comment by SpaceK33z** [comment]
@gavrushkin we could also benefit from what Seb Lee asks, to have the protocol documented. We (@ameesme and me) can't use this SDK in Modulaser because of the size and Qt dependency and are now considering to build our own Node.js package for this that directly communicates over the network.

**Comment by Grix** [comment]
I've been working on this for LaserShowGen for a bit, and made some changes to make it easier to integrate into my DAC library. It isn't tested yet, but in case someone will find it helpful the code is here: https://github.com/Grix/ildagen/tree/master/src/dacwrapper/dacwrapper

It still depends on Qt, unfortunately, but I changed all the Qt macros into [Verdigris](https://github.com/woboq/verdigris) macros, which eliminated the need for the Qt metaobject compiler, greatly simplifying the build process. I also refactored and removed a bunch of code that I didn't need, keeping only the specific wifi cube classes.

**Comment by tgreiser** [comment]
Thanks @Grix! Do you have any tips about how to set up Verdigris? I'm using your VS .sln and I see you have included wobjectdefs.h and other headers, but I am getting compiler errors and visual studio shows errors everywhere W_SIGNAL or W_SLOT appears in the code.

```
1>Device_LaserDock.cpp
1>Device_LaserDockNetwork.cpp
1>D:\Code\ildagen\src\dacwrapper\dacwrapper\LaserDockNetworkDevice.h(123,31): warning C5103: pasting '/' and '*' does not result in a valid preprocessing token
1>D:\Code\ildagen\src\dacwrapper\dacwrapper\wobjectdefs.h(629): message : in expansion of macro 'W_PARAM_TOSTRING'
1>D:\Code\ildagen\src\dacwrapper\dacwrapper\wobjectdefs.h(608): message : in expansion of macro 'W_MACRO_MSVC_EMPTY'
1>D:\Code\ildagen\src\dacwrapper\dacwrapper\wobjectdefs.h(876): message : in expansion of macro 'W_SIGNAL2'
1>D:\Code\ildagen\src\dacwrapper\dacwrapper\wobjectdefs.h(663): message : in expansion of macro 'W_RETURN'
1>D:\Code\ildagen\src\dacwrapper\dacwrapper\wobjectdefs.h(863): message : in expansion of macro 'W_SIGNAL'
1>D:\Code\ildagen\src\dacwrapper\dacwrapper\wobjectdefs.h(606): message : in expansion of macro 'W_MACRO_MSVC_EXPAND'
1>D:\Code\ildagen\src\dacwrapper\dacwrapper\LaserDockNetworkDevice.h(123,31): warning C5103: pasting '*' and '/' does not result in a valid preprocessing token
1>D:\Code\ildagen\src\dacwrapper\dacwrapper\wobjectdefs.h(629): message : in expansion of macro 'W_PARAM_TOSTRING'
1>D:\Code\ildagen\src\dacwrapper\dacwrapper\wobjectdefs.h(608): message : in expansion of macro 'W_MACRO_MSVC_EMPTY'
1>D:\Code\ildagen\src\dacwrapper\dacwrapper\wobjectdefs.h(876): message : in expansion of macro 'W_SIGNAL2'
1>D:\Code\ildagen\src\dacwrapper\dacwrapper\wobjectdefs.h(663): message : in expansion of macro 'W_RETURN'
1>D:\Code\ildagen\src\dacwrapper\dacwrapper\wobjectdefs.h(863): message : in expansion of macro 'W_SIGNAL'
1>D:\Code\ildagen\src\dacwrapper\dacwrapper\wobjectdefs.h(606): message : in expansion of macro 'W_MACRO_MSVC_EXPAND'
1>D:\Code\ildagen\src\dacwrapper\dacwrapper\LaserDockNetworkDevice.h(123,31): error C2059: syntax error: '/'
1>D:\Code\ildagen\src\dacwrapper\dacwrapper\LaserDockNetworkDevice.h(123,31): error C3553: decltype expects an expression not a type
1>D:\Code\ildagen\src\dacwrapper\dacwrapper\LaserDockNetworkDevice.h(123,31): error C2143: syntax error: missing ',' before '/'
1>D:\Code\ildagen\src\dacwrapper\dacwrapper\LaserDockNetworkDevice.h(123,31): fatal error C1903: unable to recover from previous error(s); stopping compilation
1>INTERNAL COMPILER ERROR in 'C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.26.28801\bin\HostX86\x86\CL.exe'
1>    Please choose the Technical Support command on the Visual C++
1>    Help menu, or open the Technical Support help file for more information
1>Done building project "dacwrapper.vcxproj" -- FAILED.
========== Rebuild All: 0 succeeded, 1 failed, 0 skipped ==========
```

**Comment by Grix** [comment]
@tgreiser I haven't seen those errors myself, but I think Verdigris is a bit finicky on language standards since it uses some very modern C++ macro functionality. It looks like you have VS 2019 so it should be supported, but exactly what platform toolset, C++ standard, Qt version, etc, are you using? I compiled with VS toolset v141, and Qt 5.15.2.

**Comment by ameesme** [comment]
Here's an idea; Wickedlasers should get their act together and provide a proper SDK (or protocol documentation) that can actually be worked with. If they expect other software- and hardware-engineers to support their **commercial** hardware, this is the least they could do and 100% their responsibility.

**Comment by tgreiser** [comment]
> @tgreiser I haven't seen those errors myself, but I think Verdigris is a bit finicky on language standards since it uses some very modern C++ macro functionality. It looks like you have VS 2019 so it should be supported, but exactly what platform toolset, C++ standard, Qt version, etc, are you using? I compiled with VS toolset v141, and Qt 5.15.2.

Ok, that seemed to work. Next question does your library support Windows x86 or 32 bit only?

**Comment by Grix** [comment]
> Ok, that seemed to work. Next question does your library support Windows x86 or 32 bit only?

It should support any platform in theory, but the underlying libraries need to match so you have to replace lib and dll files. In this case Qt, and libusb if you are using the rest of the library too.

**Comment by s4y** [comment]
I just got a WiFi LaserCube and I'm working on controlling it over the network. Here's a proof of concept Python program that discovers and connects to the LaserCube and draws with it:

https://gist.github.com/s4y/0675595c2ff5734e927d68caf652e3af

Feedback and suggestions are welcome. I'm likely going to continue working on this and, at the very least, make it into a simple server that exposes a WebSocket. PLEASE EXERCISE EXTREME CAUTION WORKING WITH THIS CODE! I inadvertently burned at least one object in my house early in development and so could you.

EDIT: To elaborate a little, the most important bit turned out to be keeping track of how much space is left in the laser's RX buffer and blocking when it goes below some threshold (which can be adjusted to prioritize latency or stability). Like the original implementation, I use the DAC rate to guess in advance how long it's going to take to scan out points but ultimately listen to the numbers coming back from the laser.

**Comment by sebleedelisle** [comment]
Hey @s4y nice work! I don't suppose you'd be able to add some comments to your code? It's so helpful that you've made a minimal example but I'm not particularly fluent in Python, so some additional comments would be really helpful. 

Seb

[EDIT] I should add that Wicked should really be providing documentation but in their somewhat awkward absence, I appreciate your work...

**Comment by s4y** [comment]
> I don't suppose you'd be able to add some comments to your code? It's so helpful that you've made a minimal example but I'm not particularly fluent in Python, so some additional comments would be really helpful.

@sebleedelisle Good idea. Added some, let me know what you think!

**Comment by sebleedelisle** [comment]
@s4y this looks great, how did you work all this stuff out?

**Comment by s4y** [comment]
> @s4y this looks great, how did you work all this stuff out?

@sebleedelisle Thanks! It's based on this repo, specifically these two files:

https://github.com/Wickedlasers/libLaserdockCore/blob/master/3rdparty/laserdocklib/src/LaserDockNetworkDevice.cpp
https://github.com/Wickedlasers/libLaserdockCore/blob/master/ldCore/src/Hardware/ldNetworkHardware.cpp

**Comment by SpaceK33z** [comment]
Based on @s4y's work we made a Node.js port of it for our [laser-dac repository](https://github.com/Volst/laser-dac/tree/master/packages/lasercube-wifi). Based on that we can finally add support for the LaserCube Wifi to Modulaser.

**Comment by Grix** [comment]
I scrapped my earlier attempt because even with Verdigris I couldn't get some of the Qt crap to work.
So I tried to write a driver using pure C++ sockets, with much help from @s4y 's python example. Here is the result: https://github.com/Grix/ildagen/blob/master/src/dacwrapper/dacwrapper/LaserCubeNetwork.cpp

It finally works-ish, but there are big problems, such as the output often becoming mangled and/or turning off when I send frames. I think it's a buffering issue like the remote buffer is running out of samples, even though I try to keep it a safe size. So needs more work but it's a start.

EDIT: Now it seems to work fine after some tweaks.

**Comment by paulmainwaring** [comment]
The WiFi cube can only receive a maximum of 20 UDP packets in one burst no matter how few samples are in each packet (h/w limitation).  After 20 you need a pause of ~10ms

**Comment by mattbeghin** [comment]
I agree that this SDK is unusable. I also did the work of removing all useless dependencies (audio, multimedia etc.) and only keep ldHardware & laserdocklib. But it's not easy to keep the buffer size to avoid any glitches. I'll test Grix version when I have more time. Seems to be the right thing to do. Thanks for sharing !

**Comment by gavrushkin** [comment]
Thank you all for your feedback, guys. 

We'll consider your notes and work on libLaserdockCore to make it easier for third-party developers this year. Unfortunately we don't have a lot of resources to make it easy peasy right now or even next week. So we'll work on it step by step. 

There are a lot of complaints on library complexity so the next step will be to split it into some small library and add features as modules. Later we can add versioning, more documentation, simplify API, etc. But it'll be quite slow progress. 

I won't close this issue for now but definitely will do it later so you can write your proposals in new issues one by one (or create pull requests ^^)

I've just updated this library with some features from LaserOS like lua game engine, box2d library and our own small ldRazer framework with some games as examples. Also network code is updated with important buffer size fix. So if you know lua it can be fun for you

**Comment by mattbeghin** [comment]
> I won't close this issue for now but definitely will do it later so you can write your proposals in new issues one by one (or create pull requests ^^)

Make an SDK with the smaller amount of dependencies and code as possible, with same API than laserdocklib (or libetherdream), ideally with both USB & network support (list devices, open/close, get/set dac rate, send next frame to show, wait for last frame to be transferred to device).

**Comment by charring** [comment]
Was just Googling and found this issue thread.  I am considering getting a LaserCube. Any updates in last two years?
